File: AzureIoTHubConnectedServiceLibrary/Resources/CPP/WAC/azure_iot_hub.cpp
#include <ppltasks.h>
#include <pplawait.h>
#include <stdio.h>

using namespace Microsoft::Azure::Devices::Client;
using namespace Platform;
using namespace concurrency;

//
// String containing Hostname, Device Id & Device Key in the format:
// "HostName=<host_name>;DeviceId=<device_id>;SharedAccessKey=<device_key>"
//
// Note: this connection string is specific to the device "$deviceId$". To configure other devices,
// see information on iothub-explorer at http://aka.ms/iothubgetstartedVSCS
//
static const wchar_t* connection_string = L"HostName=$iotHubUri$;DeviceId=$deviceId$;SharedAccessKey=$deviceKey$";

//
// To monitor messages sent to device "$deviceId$" use iothub-explorer as follows:
//    iothub-explorer HostName=$iotHubUri$;SharedAccessKeyName=service;SharedAccessKey=$servicePrimaryKey$ monitor-events "$deviceId$"
//

// Refer to http://aka.ms/azure-iot-hub-vs-cs-wiki for more information on Connected Service for Azure IoT Hub

task<void> send_device_to_cloud_message()
{
    auto deviceClient = DeviceClient::CreateFromConnectionString(ref new Platform::String(connection_string), TransportType::Amqp);

    std::string message = "Hello Cloud from C++ UWP App!";

    auto pbuffer = ref new Platform::Array<byte>((unsigned char*)message.data(), message.length());
    auto eventMessage = ref new Message(pbuffer);
    return create_task(deviceClient->SendEventAsync(eventMessage)).then([] {
        OutputDebugString(L"message sent successfully\n");
    });
}

#ifdef _RESUMABLE_FUNCTIONS_SUPPORTED
task<std::string> receive_cloud_to_device_message()
{
    auto deviceClient = DeviceClient::CreateFromConnectionString(ref new Platform::String(connection_string), TransportType::Amqp);

    while (true)
    {
        auto receivedMessage = await deviceClient->ReceiveAsync();

        if (receivedMessage != nullptr)
        {
            auto bytes = receivedMessage->GetBytes();
            auto messageData = bytes->Data;
            auto len = bytes->Length;

            std::string str((char*)messageData, len);

            await deviceClient->CompleteAsync(receivedMessage);

            return str;
        }

        await create_task([] { std::this_thread::sleep_for(std::chrono::seconds(1)); });
    }
}
#endif

File: AzureIoTHubConnectedServiceLibrary/Resources/CSharp/Tpm/AzureIoTHub.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.Devices.Tpm;
using Microsoft.Azure.Devices.Client;

static class AzureIoTHub
{
    //
    // This sample assumes the device has been connected to Azure with the IoT Dashboard
    //
    // Refer to http://aka.ms/azure-iot-hub-vs-cs-wiki for more information on Connected Service for Azure IoT Hub

    public static async Task SendDeviceToCloudMessageAsync()
    {
        TpmDevice myDevice = new TpmDevice($TPMSlot$); // Use logical device $TPMSlot$ on the TPM
        string hubUri = myDevice.GetHostName();
        string deviceId = myDevice.GetDeviceId();
        string sasToken = myDevice.GetSASToken();

        var deviceClient = DeviceClient.Create(
            hubUri,
            AuthenticationMethodFactory.
                CreateAuthenticationWithToken(deviceId, sasToken), TransportType.Mqtt);

#if WINDOWS_UWP
        var str = "{\"deviceId\":\"$deviceId$\",\"messageId\":1,\"text\":\"Hello, Cloud from a secure UWP C# app!\"}";
#else
        var str = "{\"deviceId\":\"$deviceId$\",\"messageId\":1,\"text\":\"Hello, Cloud from a secure C# app!\"}";
#endif
        var message = new Message(Encoding.ASCII.GetBytes(str));

        await deviceClient.SendEventAsync(message);
    }

    public static async Task<string> ReceiveCloudToDeviceMessageAsync()
    {
        TpmDevice myDevice = new TpmDevice(0); // Use logical device 0 on the TPM by default
        string hubUri = myDevice.GetHostName();
        string deviceId = myDevice.GetDeviceId();
        string sasToken = myDevice.GetSASToken();

        var deviceClient = DeviceClient.Create(
            hubUri,
            AuthenticationMethodFactory.
                CreateAuthenticationWithToken(deviceId, sasToken), TransportType.Mqtt);

        while (true)
        {
            var receivedMessage = await deviceClient.ReceiveAsync();

            if (receivedMessage != null)
            {
                var messageData = Encoding.ASCII.GetString(receivedMessage.GetBytes());
                await deviceClient.CompleteAsync(receivedMessage);
                return messageData;
            }

            await Task.Delay(TimeSpan.FromSeconds(1));
        }
    }
}
